proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+  find places where exception references are stored.
+
+Hence, while an exception reference is a new first class type, this proposal
+disallows their usage in linear memory.
+
+A WebAssembly exception is created by the `except` instruction. Once an
+exception is created, you can throw it with the `throw` instruction. Thrown
+exceptions are handled as follows:
+
+1. They can be caught by a catch block in an enclosing try block of a function
+   body. The caught exception is pushed onto the stack.
+
+1. Throws not caught within a function body continue up the call stack, popping
+   call frames, until an enclosing try block is found.
+   
+1. If the call stack is exhausted without any enclosing try blocks, the host VM
+   defines how to handle the uncaught exception.
+
+### Exceptions
+
+An `exception` is an internal construct in WebAssembly that is maintained by the
+host. WebAssembly exceptions (as opposed to host exceptions) are defined by a
+new `exception section` of a WebAssembly module. The exception section is a list
+of exception types, from which exceptions can be created.
+
+Each exception type has a `type signature`. The type signature defines the list
+of values associated with the exception.
+
+Within the module, exception types are identified by an index into the
+[exception index space](#exception-index-space). This index is referred to as
+the `exception tag`. The `tagged exception type` is the corresponding
+exception type refered to by the exception tag.
+
+Exception types can be imported and exported by adding the appropriate entries
+to the import and export sections of the module. All imported/exported exception
+types must be named to reconcile exception tags between modules.
+
+Exception tags are used by:
+
+1. The `except` instruction which creates a WebAssembly instance of the
+   corresponding tagged exception type, and pushes a reference to it onto the
+   stack.
+
+2. The `if_except` instruction that queries an exception to see if it is an
+   instance of the corresponding tagged exception class, and if true it pushes
+   the corresponding values of the exception onto the stack.
+   
+### The exception refernce data type
+
+Data types are extended to have a new `except_ref` type. The representation of
+an exception type is left to the host VM, but its size must be fixed. The actual
+number of bytes it takes to represent any `except_ref` is left to the host VM.
+
+### Try and catch blocks
+
+A _try block_ defines a list of instructions that may need to process exceptions
+and/or clean up state when an exception is thrown. Like other higher-level
+constructs, a try block begins with a `try` instruction, and ends with an `end`
+instruction. That is, a try block is sequence of instructions having the
+following form:
+
+```
+try block_type
+  instruction*
+catch
+  instruction*
+end
+```
+
+A try block ends with a `catch block` that is defined by the list of
```

Comment:
Actually, I expected layer 2 to provide a clean static model like that of the original proposal. In that context, no garbage collection is needed because the structure defines the lifetime of the exceptions, and hence it can clean up when each catch is exitted.

Reply:
@rossberg I'm not sure if I understand what you mean by the efficiency and the maximum cost of a catch-all. I think the reason we separated those two is because C++ is going to only use `catch_all`, so we need a dynamic type checking within the `catch_all`.

@KarlSchimpf So, you mean, we can use `match` and assign an exception object to a local within `catch_all` but not `catch tag` instructions that are going to be introduced in layer 2? I didn't know that... I just thought `catch tag` eliminates the need for dynamic matching in case we can separate code blocks for each catch clause, which is not feasible for C++. To me the idea that an exception object is prevented to be assigned to a local within a certain code block but not others seems a little weird. Then can we assign exception objects to locals outside of catch blocks? (not within `catch_all` or `catch tag`)

Reply:
@aheejin, yes, but C++ is an extremely degenerate case in that regard. No other language works that way. Usually you want to be able to jump to the responsible handler directly, because that is cheaper than executing various intermediate handlers only to discover that they are not relevant for the given exception.

Reply:
What I was claim wasn't on the `catch_all`, but the other catch blocks that can be added to a try block. The earlier catches (if specified) have tags and allow the embedder to optimize the matches (just like the original proposal). A separate match is needed only if you a) embed `match` in the `catch_all`, or b) you store it locally and processes it later.

Adding these catch clauses to the try block I assumed would be after MVP.